### DISCUSSION TITLE
Can I use GPU with neural-prophet?

### DISCUSSION THREAD

**Discussion by laoliang95** [ourownstory/neural_prophet#1014] [Q&A - get help using NeuralProphet] [answered]
I am searching for a long time on net. But no use. Please help or try to give some ideas how to achieve this. i want to use neuralprophet with GPU, this is my code
`trainer_config = {"accelerator":"gpu"}
    m = NeuralProphet(trainer_config=trainer_config)
    new_dt = df_ercot

    metrics = m.fit(new_dt, freq="W")`
but i got the following information and Exception
`RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!`

**Comment by noxan** [comment, marked as answer]
@laoliang95 Yes, we added gpu support in the latest release candidate and it will become available with the final `0.5.0` release. For now you can test the pre-release version either by installing from source or via pip with any version starting from `0.5.0rc1` [(link)](https://pypi.org/project/neuralprophet/0.5.0rc1/)

**Reply by laoliang95** [reply]
Thank you very much for your reply.  i will try !

**Reply by hjmok** [reply]
@laoliang95 did you get it working after changing the version to 0.5.0rc1?
I ran the following code 

`trainer_config = {"accelerator":"gpu"}`

```
m = NeuralProphet(
    weekly_seasonality=6,
    daily_seasonality=10,
    trend_reg=1,
    learning_rate=0.01,
    trainer_config=trainer_config
)
```

`df_train, df_test = m.split_df(sf_load_df, freq="H", valid_p=1.0 / 12)`

`metrics = m.fit(df_train, freq="H", validation_df=df_test, progress="plot")`

However, I still run into this error
`RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!`

I have checked that my torch is connected and currently using my NVIDIA GeForce GTX 1650 Ti from this command
`print(torch.cuda.get_device_name(0))`

**Reply by noxan** [reply]
@karl-richter is already looking into this, see https://github.com/ourownstory/neural_prophet/issues/1002

We aim to release another fix soon, glad to have your input in resolving it swiftly :)

**Reply by noxan** [reply]
@hjmok @laoliang95 - we release `0.5.0rc2` a few moments ago - it contains several bug fixes from @karl-richter for gpu support. Let us know if things are working smoothly for you now :)

**Reply by hjmok** [reply]
hi @noxan , thanks for the swift update. 
I've run into trouble when trying on different platforms. Please advise, thank you!!

1. On Google Colab, I run into the following matplotlib error after I try running NeuralProphet 0.5.0rc2, https://colab.research.google.com/drive/1Js8tnrkKUGaF3JLq_tVTCkBWlp7zIqHP?usp=sharing

```
AttributeError                            Traceback (most recent call last)
[<ipython-input-6-67f72aba51f7>](https://localhost:8080/#) in <module>
      9 df_train, df_test = m.split_df(sf_load_df, freq="H", valid_p=1.0 / 12)
     10 
---> 11 metrics = m.fit(df_train, freq="H", validation_df=df_test, progress="plot")

[/usr/local/lib/python3.8/dist-packages/neuralprophet/forecaster.py](https://localhost:8080/#) in fit(self, df, freq, validation_df, progress, minimal, continue_training)
    750             # causes an error in the Windows and MacOS environment
    751             if matplotlib.is_interactive():
--> 752                 fig.show()
    753 
    754         self.fitted = True

AttributeError: 'list' object has no attribute 'show'
```
2. When attempting on my local jupyter notebook, with Python version 3.9.12

```
Kernel Restarting
The kernel for Documents/Forecasting/neural_prophet energy.ipynb appears to have died. It will restart automatically.
```
3. When attempting to run it as a .py file in the terminal

```
OMP: Error #15: Initializing libiomp5md.dll, but found libiomp5md.dll already initialized.
OMP: Hint This means that multiple copies of the OpenMP runtime have been linked into the program. That is dangerous, since it can degrade performance or cause incorrect results. The best thing to do is to ensure that
 only a single OpenMP runtime is linked into the process, e.g. by avoiding static linking of the OpenMP runtime in any library. As an unsafe, unsupported, undocumented workaround you can set the environment variable K
MP_DUPLICATE_LIB_OK=TRUE to allow the program to continue to execute, but that may cause crashes or silently produce incorrect results. For more information, please see http://www.intel.com/software/products/support/.
```

**Reply by laoliang95** [reply]
> @laoliang95 did you get it working after changing the version to 0.5.0rc1? I ran the following code
> 
> `trainer_config = {"accelerator":"gpu"}`
> 
> ```
> m = NeuralProphet(
>     weekly_seasonality=6,
>     daily_seasonality=10,
>     trend_reg=1,
>     learning_rate=0.01,
>     trainer_config=trainer_config
> )
> ```
> 
> `df_train, df_test = m.split_df(sf_load_df, freq="H", valid_p=1.0 / 12)`
> 
> `metrics = m.fit(df_train, freq="H", validation_df=df_test, progress="plot")`
> 
> However, I still run into this error `RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!`
> 
> I have checked that my torch is connected and currently using my NVIDIA GeForce GTX 1650 Ti from this command `print(torch.cuda.get_device_name(0))`

yes. it working. i create a new project and  download the zip code from github . and run `pip install .[live]`
you can flow this document：https://github.com/ourownstory/neural_prophet/blob/main/docs/zh/%E5%BF%AB%E9%80%9F%E5%BC%80%E5%A7%8B.md

**Reply by noxan** [reply]
@hjmok There seems to be an issue with the `progress="plot"` of the `fit()` function. Do you want to report an issue on this or want me to take care of it?

For the time being best try without the `progress="plot"` option and it worked for me.
```
metrics = m.fit(df_train, freq="H", validation_df=df_test)
```


Or for your full example

```
m = NeuralProphet(
    weekly_seasonality=6,
    daily_seasonality=10,
    trend_reg=1,
    learning_rate=0.01,
    accelerator="gpu",
    epochs=4
)
df_train, df_test = m.split_df(sf_load_df, freq="H", valid_p=1.0 / 12)

metrics = m.fit(df_train, freq="H", validation_df=df_test)
```

**Comment by karl-richter** [comment]
Hi @hjmok, @laoliang95, awesome that you are already trying out the accelerator! instead of using trainer_config, you can simply use the accelerator arg of NeuralProphet in the following way:
```python
m = NeuralProphet(
    weekly_seasonality=6,
    daily_seasonality=10,
    trend_reg=1,
    learning_rate=0.01,
    accelerator="gpu"  # or "auto", "mps" for the M1 chip
)
```

**Reply by laoliang95** [reply]
tank you for your reply !

**Reply by lilGao16** [reply]
i met a problem that using gpu is slower than using cpu,wonder if it's normal or there is other parameters i can adjust,thanks
